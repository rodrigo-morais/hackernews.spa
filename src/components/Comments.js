import React from 'react'

const Comments = () => (
  <h1>Comments</h1>
)

export default Comments
