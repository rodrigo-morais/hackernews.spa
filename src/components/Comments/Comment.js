import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { HN_URL } from '../../constants'

import getTime from '../../lib/time'

import Tbody from '../shared/Tbody'
import Tr from '../shared/Tr'
import TdShared from '../shared/Td'
import AColor from '../shared/AColor'
import Vote from '../shared/Vote'

const Table = styled.table`
  padding-top: 10px;
  padding-left: 10px;
`

const Td = styled(TdShared)`
  padding: 1px;
  font-size: 8pt;
`

const TdFirst = styled(TdShared)`
  width: 1px;
`

const Commentary = styled.div`
  font-family: Verdana, Geneva, sans-serif;
  font-size: 9pt;
  color: #000000;
  max-width: 1215px;
  overflow: auto;
  margin-top: 10px;
`

const Reply = styled.div`
  color: #000000;
  font-size: x-small;
`

const AReply = styled.div`
  text-decoration: underline;
`

const Comment = ({ comment, replies }) => {
	const { time, timeMeasure } = getTime(comment.time)

  return (
    <Table>
      <Tbody>
        <Tr>
          <TdFirst>
          </TdFirst>
          <Td>
            <Vote />
          </Td>
          <Td>
            <AColor href={`${HN_URL}user?id=${comment.by}`}>{comment.by}</AColor>
            {` ${time} ${timeMeasure} ago `}
            [&nbsp;
            <AColor href="#">-</AColor>
            &nbsp;]
            <Commentary dangerouslySetInnerHTML={{ __html: comment.text }} />
            <Reply>
              <AReply href="#">reply</AReply>
            </Reply>
            { replies }
          </Td>
        </Tr>
      </Tbody>
    </Table>
  )
}

Comment.propTypes = {
	comment: PropTypes.object.isRequired,
	replies: PropTypes.array.isRequired,
}

export default Comment
