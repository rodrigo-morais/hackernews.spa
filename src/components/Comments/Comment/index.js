import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { HN_URL } from '../../../constants'

import getTime from '../../../lib/time'

import Text from './Text'
import Reply from './Reply'

import Tbody from '../../shared/Tbody'
import Tr from '../../shared/Tr'
import TdShared from '../../shared/Td'
import AColor from '../../shared/AColor'
import Vote from '../../shared/Vote'

const Table = styled.table`
  padding-top: 10px;
  padding-left: 10px;
`

const Td = styled(TdShared)`
  padding: 1px;
  font-size: 8pt;
`

const TdFirst = styled(TdShared)`
  width: 1px;
`

const Comment = ({ comment, replies }) => {
	const { time, timeMeasure } = getTime(comment.time)

  return (
    <Table>
      <Tbody>
        <Tr>
          <TdFirst>
          </TdFirst>
          <Td>
            <Vote />
          </Td>
          <Td>
            <AColor href={`${HN_URL}user?id=${comment.by}`}>{comment.by}</AColor>
            {` ${time} ${timeMeasure} ago `}
            [&nbsp;
            <AColor href="#">-</AColor>
            &nbsp;]
            <Text text={comment.text} />
            <Reply />
            { replies }
          </Td>
        </Tr>
      </Tbody>
    </Table>
  )
}

Comment.propTypes = {
	comment: PropTypes.object.isRequired,
	replies: PropTypes.array.isRequired,
}

export default Comment
