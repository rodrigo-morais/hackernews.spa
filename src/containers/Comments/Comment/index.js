import React from 'react'
import PropTypes from 'prop-types'

import CommentComponent from '../../../components/Comments/Comment'

class Comment extends React.Component {
  constructor(props) {
    super(props)

    this.toggleDetails = this.toggleDetails.bind(this)

    this.state = { isVisible: true }
  }

  toggleDetails() {
    this.setState((prevState) => ({ isVisible: !prevState.isVisible }))
  }

  render() {
    const { comment, replies } = this.props
    const { isVisible } = this.state

    return (
      <CommentComponent
        comment={comment}
        replies={replies}
        visible={isVisible}
        toggleVisibility={this.toggleDetails}
      />
    )
  }
}

Comment.propTypes = {
	comment: PropTypes.object.isRequired,
	replies: PropTypes.array.isRequired,
}

export default Comment
