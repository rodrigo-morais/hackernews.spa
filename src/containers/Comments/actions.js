import {
  REQUESTED_COMMENTS,
  RECEIVED_COMMENTS,
  FAILED_COMMENTS,
} from './constants'

export const requestComments = () => ({
  type: REQUESTED_COMMENTS,
})

export const receiveComments = (data) => ({
  type: RECEIVED_COMMENTS,
  data,
})

export const failComments = () => ({
  type: FAILED_COMMENTS,
})
