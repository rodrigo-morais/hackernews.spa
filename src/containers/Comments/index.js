import { connect } from 'react-redux'

import Comments from '../../components/Comments'

const mapStateToProps = ({ comments }) => ({ comments })

export default connect(mapStateToProps)(Comments)
