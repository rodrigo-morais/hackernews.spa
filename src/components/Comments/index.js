import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import Comment from './Comment'

import TableBody from '../shared/TableBody'
import Tbody from '../shared/Tbody'
import Tr from '../shared/Tr'

const TbodyTop = styled(Tbody)`
  padding-top: 20px;
`
const buildComment = (comment, replies) => (
  <Tr key={comment.id}>
    <td>
      <Comment comment={comment} replies={replies} />
    </td>
  </Tr>
)

const flatComments = (comments) => {
  if (comments && comments.length === 0) {
    return []
  }

  const comment = comments.pop()
  const replies = comment.kids ? flatComments(comment.kids) : []
  return [buildComment(comment, replies), ...flatComments(comments)]
}

const Comments = ({ comments: { data } }) => (
  <TableBody>
    <TbodyTop>
      { data && flatComments(data) }
    </TbodyTop>
  </TableBody>
)

Comments.propTypes = {
	comments: PropTypes.object.isRequired,
}

export default Comments
