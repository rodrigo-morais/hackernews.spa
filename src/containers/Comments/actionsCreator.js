import { HN_API } from '../../constants'
import { requestComments, receiveComments, failComments } from './actions'

const getCommentsByStoryId = (storyId) => async (dispatch) => {
  const res = await fetch(`${HN_API}item/${storyId}.json`)
  const story = await res.json()

  return getComments(story.kids)(dispatch)
}

const getComments = (commentIds) => async (dispatch) => {
  dispatch(requestComments())
  try {
    const comments = await Promise.all(commentIds.map((commentId) => getComment(commentId)))

    dispatch(receiveComments(comments))
  } catch (err) {
    dispatch(failComments())
  }
}

const getComment = (commentId) => new Promise(async (resolve, reject) => {
  try {
    const res = await fetch(`${HN_API}item/${commentId}.json`)
    const comment = await res.json()

    if (!comment.kids || comment.kids.length === 0) {
      resolve(comment)
      return
    }

    const kids = await Promise.all(comment.kids.map((reply) => getComment(reply)))
    resolve({ ...comment, ...{ kids } })
  } catch (err) {
    reject()
  }
})


export { getComments, getCommentsByStoryId }
