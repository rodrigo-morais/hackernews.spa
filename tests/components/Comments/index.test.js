import React from 'react'
import renderer from 'react-test-renderer'

import Comments from '../../../src/components/Comments'
import data from '../../data/comments'

const DATE = new Date(1537666329761)
global.Date = jest.fn(() => DATE)

const comments = { data }

describe('Comments', () => {
  describe('Rendering', () => {
    test('renders according to design', () => {
      const component = renderer.create(
				<Comments comments={comments} />
      )
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
