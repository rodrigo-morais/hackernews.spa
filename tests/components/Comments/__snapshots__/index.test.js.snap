// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments Rendering renders according to design 1`] = `
<table
  className="sc-gqjmRU aNPGS"
>
  <tbody
    className="sc-VigVT bXmzsK sc-ifAKCX bmKgXx"
  >
    <tr
      className="sc-EHOje kSzTaY"
    >
      <td>
        <table
          className="sc-dnqmqq kcpEyb"
        >
          <tbody
            className="sc-ifAKCX bmKgXx"
          >
            <tr
              className="sc-EHOje kSzTaY"
            >
              <td
                className="sc-gZMcBi lnUsYu sc-bZQynM fqVoXA"
              />
              <td
                className="sc-iwsKbI fCuVEp sc-bZQynM fqVoXA"
              >
                <center>
                  <a
                    className="sc-bwzfXH cQPqep"
                    href="https://news.ycombinator.com/vote?id=18014283&how=up&goto=news"
                  >
                    <div
                      className="sc-htoDjs fEBFom"
                    />
                  </a>
                </center>
              </td>
              <td
                className="sc-iwsKbI fCuVEp sc-bZQynM fqVoXA"
              >
                <a
                  className="sc-gzVnrw dxVyyl sc-bwzfXH cQPqep"
                  href="https://news.ycombinator.com/user?id=kizer"
                >
                  kizer
                </a>
                 0 minutess ago 
                [ 
                <a
                  className="sc-gzVnrw dxVyyl sc-bwzfXH cQPqep"
                  href="#"
                >
                  -
                </a>
                 ]
                <div
                  className="sc-bdVaJa cAiwFk"
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": Object {
                        "text": "I wonder why we&#x27;re still programming primarily in text - we ought to see more visual editors that operate on top of a source file, where changes to file sync immediately to editor, and vice-versa.<p>Also, that&#x27;s the sexiest README I&#x27;ve ever seen.",
                      },
                    }
                  }
                />
                <div
                  className="sc-htpNat dxtlPY"
                >
                  <a
                    className="sc-bxivhb gcqdII sc-bwzfXH cQPqep"
                    href="#"
                  >
                    reply
                  </a>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </td>
    </tr>
    <tr
      className="sc-EHOje kSzTaY"
    >
      <td>
        <table
          className="sc-dnqmqq kcpEyb"
        >
          <tbody
            className="sc-ifAKCX bmKgXx"
          >
            <tr
              className="sc-EHOje kSzTaY"
            >
              <td
                className="sc-gZMcBi lnUsYu sc-bZQynM fqVoXA"
              />
              <td
                className="sc-iwsKbI fCuVEp sc-bZQynM fqVoXA"
              >
                <center>
                  <a
                    className="sc-bwzfXH cQPqep"
                    href="https://news.ycombinator.com/vote?id=18014283&how=up&goto=news"
                  >
                    <div
                      className="sc-htoDjs fEBFom"
                    />
                  </a>
                </center>
              </td>
              <td
                className="sc-iwsKbI fCuVEp sc-bZQynM fqVoXA"
              >
                <a
                  className="sc-gzVnrw dxVyyl sc-bwzfXH cQPqep"
                  href="https://news.ycombinator.com/user?id=wongarsu"
                >
                  wongarsu
                </a>
                 0 minutess ago 
                [ 
                <a
                  className="sc-gzVnrw dxVyyl sc-bwzfXH cQPqep"
                  href="#"
                >
                  -
                </a>
                 ]
                <div
                  className="sc-bdVaJa cAiwFk"
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": Object {
                        "text": "The interface looks really slick, but isn&#x27;t intuitive at all. I constantly have to refer back to the video to find out how to do anything.<p>You build the graph from right to left (unusual but ok), but if you try to actually draw your lines from right to left they don&#x27;t connect, you have to draw them left to right.<p>To rename something you click the element and type the name. Fairly straightforward, except that the interface gives no indication that this is possible. No blinking text cursor or anything of the sort.<p>The &quot;arguments&quot; category actually has nothing to do with arguments, it&#x27;s just &quot;use stuff you named previously&quot;. Why is that not something that&#x27;s connected on the graph?<p>Oh, right, that&#x27;s because of the way queries&#x2F;mutations work. The name of the query is combined from the name of the query node and the name of the output node. But instead of a query having two connectors, one for input and one for output, it just has everything at a depth of one as outputs, and the child nodes of the output are the inputs (actually, that&#x27;s how types and everything else works too). I certainly wouldn&#x27;t have guessed that.<p>Compare Blender&#x27;s node editor: <a href=\\"https:&#x2F;&#x2F;code.blender.org&#x2F;wp-content&#x2F;uploads&#x2F;2012&#x2F;01&#x2F;node_group_interface18.png\\" rel=\\"nofollow\\">https:&#x2F;&#x2F;code.blender.org&#x2F;wp-content&#x2F;uploads&#x2F;2012&#x2F;01&#x2F;node_gro...</a><p>Anyone can look at a screenshot of Blender&#x27;s editor and knows what&#x27;s going on: labeled connectors, no magic with different layers, etc. Creating them is also straightforward. In comparison, if you show a graph from this tool to a random developer that works with GraphQL he&#x2F;she likely won&#x27;t understand what schema the graph is supposed to represent.",
                      },
                    }
                  }
                />
                <div
                  className="sc-htpNat dxtlPY"
                >
                  <a
                    className="sc-bxivhb gcqdII sc-bwzfXH cQPqep"
                    href="#"
                  >
                    reply
                  </a>
                </div>
                <tr
                  className="sc-EHOje kSzTaY"
                >
                  <td>
                    <table
                      className="sc-dnqmqq kcpEyb"
                    >
                      <tbody
                        className="sc-ifAKCX bmKgXx"
                      >
                        <tr
                          className="sc-EHOje kSzTaY"
                        >
                          <td
                            className="sc-gZMcBi lnUsYu sc-bZQynM fqVoXA"
                          />
                          <td
                            className="sc-iwsKbI fCuVEp sc-bZQynM fqVoXA"
                          >
                            <center>
                              <a
                                className="sc-bwzfXH cQPqep"
                                href="https://news.ycombinator.com/vote?id=18014283&how=up&goto=news"
                              >
                                <div
                                  className="sc-htoDjs fEBFom"
                                />
                              </a>
                            </center>
                          </td>
                          <td
                            className="sc-iwsKbI fCuVEp sc-bZQynM fqVoXA"
                          >
                            <a
                              className="sc-gzVnrw dxVyyl sc-bwzfXH cQPqep"
                              href="https://news.ycombinator.com/user?id=aexol"
                            >
                              aexol
                            </a>
                             0 minutess ago 
                            [ 
                            <a
                              className="sc-gzVnrw dxVyyl sc-bwzfXH cQPqep"
                              href="#"
                            >
                              -
                            </a>
                             ]
                            <div
                              className="sc-bdVaJa cAiwFk"
                              dangerouslySetInnerHTML={
                                Object {
                                  "__html": Object {
                                    "text": "Thanks for that feedback. This is really valuable for me. I will correct everything you described here. I am not an UX designer to, so maybe this is the reason of bad UX and UI.<p>- I will correct drawing the lines
- I will add a blinking cursor during renaming of the node
- I will correct the query node to have input and output<p>So definitely there is lot of work to do. Deadline for these 3 - Thursday.<p>I used to work for Visual Effects industry 7 years ago. I was a TD and python scripts programmer. That&#x27;s why you have spacebar menu ( similar to this in Maya - &quot;Hotbox&quot; ).<p>Your feedback is so constructive, THANK YOU Man.",
                                  },
                                }
                              }
                            />
                            <div
                              className="sc-htpNat dxtlPY"
                            >
                              <a
                                className="sc-bxivhb gcqdII sc-bwzfXH cQPqep"
                                href="#"
                              >
                                reply
                              </a>
                            </div>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </td>
                </tr>
              </td>
            </tr>
          </tbody>
        </table>
      </td>
    </tr>
  </tbody>
</table>
`;
